vowpalwabbit/arguments.h

```diff
@@ -0,0 +1,157 @@
+    }
+
+    static size_t type_hash() {
+      return typeid(T).hash_code();
+    }
+
+    typed_argument& default_value(T value) {
+      m_default_value = std::make_shared<T>(value);
+      return *this;
+    }
+
+    bool default_value_supplied() {
+      return m_default_value.get() != nullptr;
+    }
+
+    T default_value() {
+      return m_default_value ? *m_default_value : T();
+    }
+
+    typed_argument& short_name(std::string short_name) {
+      m_short_name = short_name;
+      return *this;
+    }
+
+    typed_argument& help(std::string help) {
+      m_help = help; return *this;
+    }
+
+    typed_argument& keep(bool keep = true) {
+      m_keep = keep; return *this;
+    }
+
+    bool value_supplied() {
+      return m_value.get() != nullptr;
+    }
+
+    typed_argument& value(T& value) {
+      m_value = std::make_shared<T>(value);
+      return *this;
+    }
+
+    T value(){
+      return m_value ? *m_value : T();
+    }
+
+    std::vector<T*> m_locations;
+
+  private:
+    // Would prefer to use std::optional (C++17) here but we are targeting C++11
+    std::shared_ptr<T> m_value{ nullptr };
+    std::shared_ptr<T> m_default_value{ nullptr };
+  };
+
+  template<typename T>
+  typed_argument<T> make_typed_arg(std::string name, T* location) {
+    return typed_argument<T>(name, location);
+  }
+
+  struct argument_group_definition {
+    argument_group_definition(std::string name)
+      : m_name(name)
+    {}
+
+    template<typename T>
+    void add(typed_argument<T> op) {
+      m_arguments.push_back(std::make_shared<typed_argument<T>>(op));
+    }
+
+    template<typename T>
+    argument_group_definition& operator()(typed_argument<T> op) {
+      add(op);
+      return *this;
+    }
+
+    std::string m_name;
+    std::vector<std::shared_ptr<base_argument>> m_arguments;
+  };
+
+  struct arguments_i {
+    virtual void add_and_parse(argument_group_definition group) = 0;
+    virtual bool was_supplied(std::string key) = 0;
+    virtual std::string help() = 0;
+
+    virtual void merge(arguments_i* other) = 0;
+
+    virtual std::vector<std::shared_ptr<base_argument>>& get_all_args() = 0;
+    virtual base_argument& get_arg(std::string key) = 0;
+
+    template <typename T>
+    typed_argument<T>& get_typed_arg(std::string key) {
+      base_argument& base = get_arg(key);
+      if (base.m_type_hash != typed_argument<T>::type_hash()) {
+        throw std::bad_cast();
+      }
+
+      return dynamic_cast<typed_argument<T>&>(base);
+    }
+
+    // Will throw if any options were supplied that do not having a matching argument specification.
+    virtual void check_unregistered() = 0;
+
+    virtual ~arguments_i() {}
+  };
+
+  struct arguments_serializer_i {
+    virtual void add(base_argument& argument) = 0;
+    virtual std::string str() = 0;
+    virtual const char* data() = 0;
+    virtual size_t size() = 0;
+  };
+
+  template<typename T>
+  bool operator==(typed_argument<T>& lhs, typed_argument<T>& rhs) {

```

Comment:
Why isn't this a member of typed_argument<T>?

Reply:
There doesn't exist a need for it to be a member function and keeping it free like this is better encapsulated. http://www.drdobbs.com/cpp/how-non-member-functions-improve-encapsu/184401197?pgno=1

Reply:
Interesting.  

It seems like there are two things to worry about: 
(1) Polluting a class with functions reduces encapsulation of the data in the class.  This is a good point. 
(2) Polluting the top level namespace with function names.   I've been bitten by this a few times and it's tricky to debug because some binary you don't expect is getting linked.  

It seems like (1) and (2) can be dealt with by having a namespace for arguments? (Or should that be options for semantic clarity?)

Reply:
I 100% agree, a namespace for all of these argument related things would be good. Perhaps `config` or `configuration` to indicate the purpose of the classes

Reply:
Ok, let's do that.